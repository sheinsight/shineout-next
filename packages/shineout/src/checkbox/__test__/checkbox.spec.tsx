import { render, cleanup, fireEvent, screen } from '@testing-library/react';
import Checkbox from '..';
import mountTest from '../../tests/mountTest';
import structureTest, { inputTest, classLengthTest } from '../../tests/structureTest';
import disabledTest, { componentProps } from '../../tests/disabledTest';
import {
  classTest,
  snapshotTest,
  hasAttributesTest,
  attributesTest,
  textContentTest,
  childrenTest,
  baseTest,
  displayTest,
  createClassName,
} from '../../tests/utils';
import CheckboxBase from '../__example__/001-base';
import CheckboxStatus from '../__example__/002-checked-1';
import CheckboxIndeterminate from '../__example__/002-checked-2';
import CheckboxHtmlValue from '../__example__/003-value';
// import CheckboxInputable from '../__example__/008-inputable';
import CheckboxClick from '../__example__/009-click';

const SO_PREFIX = 'checkbox';
const originClasses = ['wrapper', 'input', 'group', 'desc'];
const originItemClasses = ['wrapperChecked', 'wrapperDisabled', 'wrapperIndeterminate'];
const {
  wrapper: checkboxClassName,
  wrapperChecked: checkboxCheckedClassName,
  wrapperDisabled: checkboxDisabledClassName,
  wrapperIndeterminate: checkboxIndeterminateClassName,
  input: checkboxInputClassName,
  group: checkboxGroupClassName,
  desc,
} = createClassName(SO_PREFIX, originClasses, originItemClasses);

const attributes = [
  {
    attribute: 'input',
    num: 1,
  },
  {
    attribute: 'i',
    num: 1,
  },
  {
    attribute: 'span',
    num: 2,
  },
];

const checkTest = (container: HTMLElement, status: boolean) => {
  const checkbox = container.querySelector(checkboxClassName)!;
  classTest(checkbox, checkboxCheckedClassName, status);
  const input = checkbox.querySelector('input')!;
  hasAttributesTest(input, 'checked', status);
  fireEvent.click(checkbox);
  classTest(checkbox, checkboxCheckedClassName, status);
  hasAttributesTest(input, 'checked', status);
};

afterEach(cleanup);
describe('Checkbox[Base]', () => {
  mountTest(Checkbox);
  baseTest(Checkbox, checkboxClassName);
  structureTest(<Checkbox>Checkbox</Checkbox>, attributes);
  childrenTest(Checkbox, checkboxClassName);
  displayTest(Checkbox, 'ShineoutCheckbox');
  snapshotTest(<CheckboxBase />);
  test('should render about class when not set attribute', () => {
    const renderContent = <Checkbox>Checkbox</Checkbox>;
    const inputAttributes = [
      {
        attribute: 'style',
        value: 'display: none;',
      },
      {
        attribute: 'tabindex',
        value: '0',
      },
      {
        attribute: 'type',
        value: 'checkbox',
      },
    ];
    inputTest(renderContent, inputAttributes);
    const { container } = render(renderContent);
    attributesTest(container.querySelector('i')!, 'tabindex', '1');
    screen.debug();
    textContentTest(container.querySelector(desc)!, 'Checkbox');
  });
  test('should render when click', () => {
    const clickFn = jest.fn();
    const { container } = render(<Checkbox onClick={clickFn}>Checkbox</Checkbox>);
    const checkbox = container.querySelector(checkboxClassName)!;
    classTest(checkbox, checkboxCheckedClassName, false);
    fireEvent.click(checkbox);
    classTest(checkbox, checkboxCheckedClassName);
    expect(clickFn.mock.calls.length).toBe(1);
    fireEvent.click(checkbox);
    classTest(checkbox, checkboxCheckedClassName, false);
    expect(clickFn.mock.calls.length).toBe(2);
  });
});
describe('Checkbox[Checked, disabled]', () => {
  snapshotTest(<CheckboxStatus />, 'about checked');
  snapshotTest(<CheckboxIndeterminate />, 'about checked is indeterminate');
  test('should set correct class', () => {
    const { container } = render(<CheckboxStatus />);
    // array[0]: checked (false: -1, 'indeterminate: 0', true: 1)
    // array[1]: disabled (false: 0, true: 1)
    const statusMap = [
      [-1, 0],
      [1, 0],
      [0, 0],
      [-1, 1],
      [1, 1],
      [0, 1],
    ];
    container.querySelectorAll(checkboxClassName).forEach((checkbox, index) => {
      classTest(checkbox, checkboxCheckedClassName, statusMap[index][0] === 1 ? true : false);
      const input = checkbox.querySelector('input')!;
      hasAttributesTest(input, 'checked', statusMap[index][0] === 1 ? true : false);
      classTest(checkbox, checkboxIndeterminateClassName, statusMap[index][0] === 0 ? true : false);
      classTest(checkbox, checkboxDisabledClassName, !!statusMap[index][1]);
      hasAttributesTest(input, 'disabled', !!statusMap[index][1]);
    });
  });
  test('should render when set checked', () => {
    const clickFn = jest.fn();
    const { container } = render(
      <Checkbox onClick={clickFn} checked>
        Checkbox
      </Checkbox>,
    );
    checkTest(container, true);
    expect(clickFn.mock.calls.length).toBe(1);
  });
  test('should render when set checked is false', () => {
    const clickFn = jest.fn();
    const { container } = render(
      <Checkbox onClick={clickFn} checked={false}>
        Checkbox
      </Checkbox>,
    );
    checkTest(container, false);
    expect(clickFn.mock.calls.length).toBe(1);
  });
  test('should render when set checked is indeterminate', () => {
    const clickFn = jest.fn();
    const { container } = render(
      <Checkbox onClick={clickFn} checked='indeterminate'>
        Checkbox
      </Checkbox>,
    );
    checkTest(container, false);
    expect(clickFn.mock.calls.length).toBe(1);
  });
  test('should render about indetermimnate', () => {
    const { container } = render(<CheckboxIndeterminate />);
    const checkboxAll = container.querySelectorAll(checkboxClassName)[0];
    const checkboxs = container
      .querySelector(checkboxGroupClassName)
      ?.querySelectorAll(checkboxClassName) as NodeListOf<Element>;
    const checkboxAllTest = (iStatus: boolean, cStatus: boolean) => {
      classTest(checkboxAll, checkboxIndeterminateClassName, iStatus);
      classTest(checkboxAll, checkboxCheckedClassName, cStatus);
    };
    expect(checkboxAll.textContent).toBe('CheckAll');
    checkboxAllTest(false, false);
    fireEvent.click(checkboxs[0]);
    expect(checkboxAll.textContent).toBe('CheckAll');
    checkboxAllTest(true, false);
    classTest(checkboxs[0], checkboxCheckedClassName);
    fireEvent.click(checkboxs[1]);
    fireEvent.click(checkboxs[2]);
    expect(checkboxAll.textContent).toBe('UnCheckAll');
    checkboxAllTest(false, true);
    fireEvent.click(checkboxs[0]);
    expect(checkboxAll.textContent).toBe('CheckAll');
    checkboxAllTest(true, false);
    classTest(checkboxs[0], checkboxCheckedClassName, false);
  });

  test('should render when set disabled', () => {
    disabledTest(Checkbox as React.FC<componentProps>, checkboxClassName, checkboxCheckedClassName);
  });
});
describe('Checkbox[HtmlValue]', () => {
  snapshotTest(<CheckboxHtmlValue />, 'about htmlValue');
  test('should renturn htmlValue', () => {
    const clickFn = jest.fn();
    const changeFn = jest.fn();
    const checkedText = 'ok';
    const { container } = render(
      <Checkbox htmlValue={checkedText} onClick={clickFn} onChange={changeFn}>
        Checkbox
      </Checkbox>,
    );
    const checkbox = container.querySelector(checkboxClassName)!;
    fireEvent.click(checkbox);
    expect(changeFn.mock.calls[0][0]).toBe(checkedText);
  });
});
describe('Checkbox[Inputable]', () => {
  // snapshotTest(<CheckboxInputable />, 'about inputable');
  test('should show input while selected', () => {
    const clickFn = jest.fn();
    const { container } = render(
      <Checkbox inputable onClick={clickFn}>
        more...
      </Checkbox>,
    );
    classLengthTest(container, checkboxInputClassName, 0);
    const checkbox = container.querySelector(checkboxClassName)!;
    fireEvent.click(checkbox);

    expect(clickFn.mock.calls.length).toBe(1);
    const input = container.querySelector(checkboxInputClassName)!;
    classLengthTest(container, checkboxInputClassName, 1);
    classLengthTest(input, 'input', 1);
    attributesTest(input?.querySelector('input') as Element, 'value', '');

    fireEvent.change(input?.querySelector('input') as HTMLInputElement, {
      target: {
        value: 'no',
      },
    });
    attributesTest(input?.querySelector('input') as Element, 'value', 'no');
  });
  test('should render while selected by htmlValue', () => {
    const { container } = render(
      <Checkbox inputable htmlValue='no'>
        more...
      </Checkbox>,
    );
    const checkbox = container.querySelector(checkboxClassName)!;
    fireEvent.click(checkbox);
    expect(
      container
        .querySelector(checkboxInputClassName)
        ?.querySelector('input')
        ?.getAttribute('value'),
    ).not.toBe('no');
  });
});
describe('Checkbox[Value]', () => {
  test('should checked same as value === htmlValue', () => {
    const clickFn = jest.fn();
    const { container } = render(<Checkbox htmlValue='red' value='blue' onClick={clickFn} />);
    const checkbox = container.querySelector(checkboxClassName)!;
    fireEvent.click(checkbox);
    classTest(checkbox, checkboxCheckedClassName, false);
  });
});
describe('Checkbox[Click]', () => {
  snapshotTest(<CheckboxClick />, 'about click');
  test('should render when click', () => {
    const { container } = render(<CheckboxClick />);
    const checkbox = container.querySelector(checkboxClassName)!;
    textContentTest(checkbox, 'Click Me 0 Times!');
    fireEvent.click(checkbox);
    textContentTest(checkbox, 'Click Me 1 Times!');
    fireEvent.click(checkbox);
    textContentTest(checkbox, 'Click Me 2 Times!');
  });
});
