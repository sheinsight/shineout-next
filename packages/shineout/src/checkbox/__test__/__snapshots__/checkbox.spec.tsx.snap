// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox[Base] should render correctly  1`] = `
<div
  class="so-checkbox-wrapper"
>
  <input
    style="display: none;"
    tabindex="0"
    type="checkbox"
  />
  <span
    class="so-checkbox-indicator-wrapper"
  >
    <i
      class="so-checkbox-indicator"
      tabindex="1"
    />
  </span>
  <span
    class="so-checkbox-desc"
  >
    Checkbox
  </span>
</div>
`;

exports[`Checkbox[Checked, disabled] should render correctly about checked 1`] = `
<div>
  <div
    class="so-checkbox-wrapper"
  >
    <input
      style="display: none;"
      tabindex="0"
      type="checkbox"
    />
    <span
      class="so-checkbox-indicator-wrapper"
    >
      <i
        class="so-checkbox-indicator"
        tabindex="1"
      />
    </span>
    <span
      class="so-checkbox-desc"
    >
      not checked
    </span>
  </div>
  <div
    class="so-checkbox-wrapper so-checkbox-wrapper-checked"
  >
    <input
      checked=""
      style="display: none;"
      tabindex="0"
      type="checkbox"
    />
    <span
      class="so-checkbox-indicator-wrapper"
    >
      <i
        class="so-checkbox-indicator"
        tabindex="1"
      >
        <svg
          fill="currentColor"
          viewBox="0 0 16 16"
        >
          <path
            d="M2.2216 7.27815C2.61213 6.88763 3.24529 6.88763 3.63582 7.27815L7.17135 10.8137C7.56188 11.2042 7.56188 11.8374 7.17135 12.2279V12.2279C6.78083 12.6184 6.14766 12.6184 5.75714 12.2279L2.2216 8.69237C1.83108 8.30184 1.83108 7.66868 2.2216 7.27815V7.27815Z"
          />
          <path
            d="M14.2431 3.7426C14.6336 4.13312 14.6336 4.76629 14.2431 5.15681L7.17202 12.2279C6.7815 12.6184 6.14833 12.6184 5.75781 12.2279V12.2279C5.36728 11.8374 5.36728 11.2042 5.75781 10.8137L12.8289 3.7426C13.2194 3.35208 13.8526 3.35208 14.2431 3.7426V3.7426Z"
          />
        </svg>
      </i>
    </span>
    <span
      class="so-checkbox-desc"
    >
      checked
    </span>
  </div>
  <div
    class="so-checkbox-wrapper so-checkbox-wrapper-indeterminate"
  >
    <input
      style="display: none;"
      tabindex="0"
      type="checkbox"
    />
    <span
      class="so-checkbox-indicator-wrapper"
    >
      <i
        class="so-checkbox-indicator"
        tabindex="1"
      >
        <svg
          fill="currentColor"
          viewBox="0 0 16 16"
        >
          <path
            d="M3 8C3 7.44772 3.44772 7 4 7H12C12.5523 7 13 7.44772 13 8V8C13 8.55228 12.5523 9 12 9H4C3.44772 9 3 8.55228 3 8V8Z"
          />
        </svg>
      </i>
    </span>
    <span
      class="so-checkbox-desc"
    >
      indeterminate
    </span>
  </div>
  <div
    class="so-checkbox-wrapper so-checkbox-wrapper-disabled"
  >
    <input
      disabled=""
      style="display: none;"
      tabindex="0"
      type="checkbox"
    />
    <span
      class="so-checkbox-indicator-wrapper"
    >
      <i
        class="so-checkbox-indicator"
        tabindex="-1"
      />
    </span>
    <span
      class="so-checkbox-desc"
    >
      not checked
    </span>
  </div>
  <div
    class="so-checkbox-wrapper so-checkbox-wrapper-disabled so-checkbox-wrapper-checked"
  >
    <input
      checked=""
      disabled=""
      style="display: none;"
      tabindex="0"
      type="checkbox"
    />
    <span
      class="so-checkbox-indicator-wrapper"
    >
      <i
        class="so-checkbox-indicator"
        tabindex="-1"
      >
        <svg
          fill="currentColor"
          viewBox="0 0 16 16"
        >
          <path
            d="M2.2216 7.27815C2.61213 6.88763 3.24529 6.88763 3.63582 7.27815L7.17135 10.8137C7.56188 11.2042 7.56188 11.8374 7.17135 12.2279V12.2279C6.78083 12.6184 6.14766 12.6184 5.75714 12.2279L2.2216 8.69237C1.83108 8.30184 1.83108 7.66868 2.2216 7.27815V7.27815Z"
          />
          <path
            d="M14.2431 3.7426C14.6336 4.13312 14.6336 4.76629 14.2431 5.15681L7.17202 12.2279C6.7815 12.6184 6.14833 12.6184 5.75781 12.2279V12.2279C5.36728 11.8374 5.36728 11.2042 5.75781 10.8137L12.8289 3.7426C13.2194 3.35208 13.8526 3.35208 14.2431 3.7426V3.7426Z"
          />
        </svg>
      </i>
    </span>
    <span
      class="so-checkbox-desc"
    >
      checked
    </span>
  </div>
  <div
    class="so-checkbox-wrapper so-checkbox-wrapper-disabled so-checkbox-wrapper-indeterminate"
  >
    <input
      disabled=""
      style="display: none;"
      tabindex="0"
      type="checkbox"
    />
    <span
      class="so-checkbox-indicator-wrapper"
    >
      <i
        class="so-checkbox-indicator"
        tabindex="-1"
      >
        <svg
          fill="currentColor"
          viewBox="0 0 16 16"
        >
          <path
            d="M3 8C3 7.44772 3.44772 7 4 7H12C12.5523 7 13 7.44772 13 8V8C13 8.55228 12.5523 9 12 9H4C3.44772 9 3 8.55228 3 8V8Z"
          />
        </svg>
      </i>
    </span>
    <span
      class="so-checkbox-desc"
    >
      indeterminate
    </span>
  </div>
</div>
`;

exports[`Checkbox[Checked, disabled] should render correctly about checked is indeterminate 1`] = `
<div>
  <div
    class="so-checkbox-wrapper"
  >
    <input
      style="display: none;"
      tabindex="0"
      type="checkbox"
    />
    <span
      class="so-checkbox-indicator-wrapper"
    >
      <i
        class="so-checkbox-indicator"
        tabindex="1"
      />
    </span>
    <span
      class="so-checkbox-desc"
    >
      CheckAll
    </span>
  </div>
  <div
    class="so-checkbox-group"
    style="margin-top: 12px;"
  >
    <div
      class="so-checkbox-wrapper"
    >
      <input
        style="display: none;"
        tabindex="0"
        type="checkbox"
      />
      <span
        class="so-checkbox-indicator-wrapper"
      >
        <i
          class="so-checkbox-indicator"
          tabindex="1"
        />
      </span>
      <span
        class="so-checkbox-desc"
      >
        Option1
      </span>
    </div>
    <div
      class="so-checkbox-wrapper"
    >
      <input
        style="display: none;"
        tabindex="0"
        type="checkbox"
      />
      <span
        class="so-checkbox-indicator-wrapper"
      >
        <i
          class="so-checkbox-indicator"
          tabindex="1"
        />
      </span>
      <span
        class="so-checkbox-desc"
      >
        Option2
      </span>
    </div>
    <div
      class="so-checkbox-wrapper"
    >
      <input
        style="display: none;"
        tabindex="0"
        type="checkbox"
      />
      <span
        class="so-checkbox-indicator-wrapper"
      >
        <i
          class="so-checkbox-indicator"
          tabindex="1"
        />
      </span>
      <span
        class="so-checkbox-desc"
      >
        Option3
      </span>
    </div>
  </div>
</div>
`;

exports[`Checkbox[Click] should render correctly about click 1`] = `
<div>
  <div
    class="so-checkbox-wrapper"
  >
    <input
      style="display: none;"
      tabindex="0"
      type="checkbox"
    />
    <span
      class="so-checkbox-indicator-wrapper"
    >
      <i
        class="so-checkbox-indicator"
        tabindex="1"
      />
    </span>
    <span
      class="so-checkbox-desc"
    >
      Click Me
       0 Times!
    </span>
  </div>
</div>
`;

exports[`Checkbox[HtmlValue] should render correctly about htmlValue 1`] = `
<div
  class="so-checkbox-wrapper"
>
  <input
    style="display: none;"
    tabindex="0"
    type="checkbox"
  />
  <span
    class="so-checkbox-indicator-wrapper"
  >
    <i
      class="so-checkbox-indicator"
      tabindex="1"
    />
  </span>
  <span
    class="so-checkbox-desc"
  >
    value is "ok"
  </span>
</div>
`;
